extensions/opentelemetry/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/tracing/reactivemessaging/ReactiveMessagingTracingDecorator.java

```diff
@@ -0,0 +1,83 @@
+package io.quarkus.opentelemetry.runtime.tracing.reactivemessaging;
+
+import java.util.List;
+
+import javax.enterprise.context.ApplicationScoped;
+
+import org.eclipse.microprofile.reactive.messaging.Message;
+
+import io.opentelemetry.context.Context;
+import io.quarkus.opentelemetry.runtime.QuarkusContextStorage;
+import io.smallrye.mutiny.Multi;
+import io.smallrye.reactive.messaging.PublisherDecorator;
+import io.smallrye.reactive.messaging.SubscriberDecorator;
+import io.smallrye.reactive.messaging.TracingMetadata;
+import io.smallrye.reactive.messaging.providers.locals.LocalContextMetadata;
+
+/**
+ * Intercepts incoming and outgoing messages from Reactive Messaging connectors.
+ * <p>
+ * For incoming messages, it fetches OpenTelemetry context from the message and attaches to the duplicated context of the
+ * message.
+ * Consumer methods will be called on this duplicated context, so the OpenTelemetry context associated with the incoming message
+ * will be propagated.
+ * <p>
+ * For outgoing messages, if the message doesn't already contain a tracing metadata, it attaches one with the current
+ * OpenTelemetry context.
+ * Reactive messaging outbound connectors, if tracing is supported, will use that context as parent span to trace outbound
+ * message transmission.
+ */
+@ApplicationScoped
+public class ReactiveMessagingTracingDecorator implements PublisherDecorator, SubscriberDecorator {
+
+    @Override
+    public int getPriority() {
+        return 1000;
+    }
+
+    /**
+     * Incoming messages
+     */
+    @Override
+    public Multi<? extends Message<?>> decorate(Multi<? extends Message<?>> publisher,
+            String channelName, boolean isConnector) {
+        Multi<? extends Message<?>> multi = publisher;
+        if (isConnector) {
+            // attach OTel context from incoming message to the duplicated message context
+            multi = multi.invoke(m -> {
+                var messageContext = m.getMetadata(LocalContextMetadata.class)
+                        .map(LocalContextMetadata::context)
+                        .orElse(null);
+                var otelContext = TracingMetadata.fromMessage(m)
```

Comment:
This will insert a loot of data into the message. 
Using a propagator that only sends the W3C headers, would optimal: https://www.w3.org/TR/trace-context/#relationship-between-the-headers 
This can be achieved with an instrumenter.

Reply:
This doesn't insert any data into the message. This part is for incoming messages, it attaches the incoming message's Otel context to the duplicated context on which the message consumer method will run. 

On the outgoing side, the other method in this class, we add tracing metadata to the outgoing message, for the instrumenter in the reactive messaging side to pick it as parent and use it for the producer trace.

Reply:
makes sense to me now. Thanks Ozan!